Documentation/design-docs/object-stack-allocation.md

```diff
@@ -0,0 +1,190 @@
+# Object Stack Allocation
+
+This document describes work to enable object stack allocation in .NET Core.
+
+## Motivation
+
+In .NET instances of object types are allocated on the garbage-collected heap.
+Such allocations have performance overhead at garbage collection time. The allocator also has to ensure that the memory is fully zero-initialized.
+If the lifetime of an object is bounded by the lifetime of the allocating method, the allocation
+may be moved to the stack. The benefits of this optimization:
+
+* The pressure on the garbage collector is reduced because the GC heap becomes smaller.
```

Comment:
This seems to miss a key point that the GC doesn't have to be involved in allocating or deallocating these objects. You mention that above, but you repeat the zero-initialize benefit below, so it might be good to clarify this.

Reply:
Clarified this in the doc.